packages/astro/src/vite-plugin-astro-server/plugin.ts

```diff
@@ -51,13 +54,33 @@ export default function createVitePluginAstroServer({
 			const localStorage = new AsyncLocalStorage();
 
 			/** rebuild the route cache + manifest */
-			async function rebuildManifest() {
+			async function rebuildManifest(path: string | null = null) {
 				pipeline.clearRouteCache();
-				routeManifest = injectDefaultDevRoutes(
-					settings,
-					devSSRManifest,
-					await createRouteManifest({ settings, fsMod }, logger), // TODO: Handle partial updates to the manifest
-				);
+
+				// If a route changes, we check if it's part of the manifest and check for its prerender value
+				if (path !== null) {
+					const route = routeManifest.routes.find((r) => path.endsWith(r.component));
+					if (!route) {
+						return;
+					}
+					if (route.type !== 'page' && route.type !== 'endpoint') return;
+
+					const routePath = fileURLToPath(new URL(route.component, settings.config.root));
+					if (!fsMod.existsSync(routePath)) {
+						// Route was renamed so it does not exist anymore
+						return;
+					}
+					const content = await fsMod.promises.readFile(routePath, 'utf-8');
```

Comment:
There is technically a race here. The file could be deleted in between the call to `existsSync` and `readFile`, this would throw an error in that case. We can change this to use only `readFile` and detect a non-existing file from the error.